glommio/src/sys/mod.rs

```diff
@@ -558,3 +573,163 @@ pub(crate) fn shutdown(raw: RawFd, how: Shutdown) -> io::Result<()> {
+    pub files_closed: u64,
+    pub file_reads: u64,
+    pub file_bytes_read: u64,
+    pub file_buffered_reads: u64,
+    pub file_buffered_bytes_read: u64,
+    pub file_writes: u64,
+    pub file_bytes_written: u64,
+    pub file_buffered_writes: u64,
+    pub file_buffered_bytes_written: u64,
+}
+
+impl RingIoStats {
+    fn new() -> Self {
+        Default::default()
+    }
+
+    /// The total amount of files opened in this executor so far.
+    ///
+    /// [files_opened] - [files_closed] gives the current open files count
+    pub fn files_opened(&self) -> u64 {
+        self.files_opened
+    }
+
+    /// The total amount of files closed in this executor so far.
+    ///
+    /// [files_opened] - [files_closed] gives the current open files count
+    pub fn files_closed(&self) -> u64 {
+        self.files_opened
+    }
+
+    /// File read IO stats
+    ///
+    /// Returns the number of individual read ops as well as bytes read
+    pub fn file_reads(&self) -> (u64, u64) {
+        (self.file_reads, self.file_bytes_read)
+    }
+
+    /// Buffered file read IO stats
+    ///
+    /// Returns the number of individual buffered read ops as well as bytes read
+    pub fn file_buffered_reads(&self) -> (u64, u64) {
+        (self.file_buffered_reads, self.file_buffered_bytes_read)
+    }
+
+    /// File write IO stats
+    ///
+    /// Returns the number of individual write ops as well as bytes written
+    pub fn file_writes(&self) -> (u64, u64) {
+        (self.file_writes, self.file_bytes_written)
+    }
+
+    /// Buffered file write IO stats
+    ///
+    /// Returns the number of individual buffered write ops as well as bytes
+    /// written
+    pub fn file_buffered_writes(&self) -> (u64, u64) {
+        (self.file_buffered_writes, self.file_buffered_bytes_written)
+    }
+}
+
+impl Sum<RingIoStats> for RingIoStats {
+    fn sum<I: Iterator<Item = RingIoStats>>(iter: I) -> Self {
+        iter.fold(RingIoStats::new(), |a, b| RingIoStats {
+            files_opened: a.files_opened + b.files_opened,
+            files_closed: a.files_closed + b.files_closed,
+            file_reads: a.file_reads + b.file_reads,
+            file_bytes_read: a.file_bytes_read + b.file_bytes_read,
+            file_buffered_reads: a.file_buffered_reads + b.file_buffered_reads,
+            file_buffered_bytes_read: a.file_buffered_bytes_read + b.file_buffered_bytes_read,
+            file_writes: a.file_writes + b.file_writes,
+            file_bytes_written: a.file_bytes_written + b.file_bytes_written,
+            file_buffered_writes: a.file_buffered_writes + b.file_buffered_writes,
+            file_buffered_bytes_written: a.file_buffered_bytes_written
+                + b.file_buffered_bytes_written,
+        })
+    }
+}
+
+/// Stores information about IO
+#[derive(Debug)]
+pub struct IoStats {
+    pub main_ring: RingIoStats,
+    pub latency_ring: RingIoStats,
+    pub poll_ring: RingIoStats,
+    pub all_rings: RingIoStats,
```

Comment:
why do we need an extra field for all rings?

Reply:
it's provided for convenience (and also so that we map-reduce only once). Do you want me to replace it with a method that compute that on-demand?

Reply:
yes, that's better

Reply:
and then maybe we don't need the other methods on `Iostats`